src/tribler-core/tribler_core/components/tag/rules/tag_rules_processor.py

```diff
@@ -0,0 +1,109 @@
+        Default values for batch_size and interval are chosen so that tag processing is not too heavy
+        fot CPU and with this values 360k items will be processed within the hour.
+        """
+        super().__init__()
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        self.notifier = notifier
+        self.db = db
+        self.mds = mds
+        self.batch_size = batch_size
+        self.interval = interval
+        self.notifier.add_observer(torrent_metadata.NEW_TORRENT_METADATA_CREATED,
+                                   callback=self.process_torrent_title)
+        self.register_task(name=self.process_batch.__name__,
+                           interval=interval,
+                           task=self.process_batch)
+
+    @db_session
+    def process_batch(self) -> int:
+        def query(_start, _end):
+            return lambda t: _start < t.rowid and t.rowid <= _end and \
+                             t.metadata_type == REGULAR_TORRENT and \
+                             t.tag_version < self.version
+
+        start = int(self.mds.get_value(LAST_PROCESSED_TORRENT_ID, default='0'))
+        end = start + self.batch_size
+        self.logger.info(f'Processing batch [{start}...{end}]')
+
+        batch = self.mds.TorrentMetadata.select(query(start, end))
+        processed = 0
+        added = 0
+        for torrent in batch:
+            added += self.process_torrent_title(torrent.infohash, torrent.title)
+            torrent.tag_version = self.version
+            processed += 1
+
+        self.logger.info(f'Processed: {processed} titles. Added {added} tags.')
+        max_row_id = self.mds.get_max_rowid()
+
+        is_beyond_the_boundary = end > max_row_id
+        if is_beyond_the_boundary:
+            self._schedule_new_process_batch_round()
+        else:
+            self.mds.set_value(LAST_PROCESSED_TORRENT_ID, str(end))
+        return processed
+
+    def process_torrent_title(self, infohash: Optional[bytes] = None, title: Optional[str] = None) -> int:
+        if not infohash or not title:
+            return 0
+        tags = set(extract_only_valid_tags(title))
+        if tags:
+            self.save_tags(infohash, tags)
+        return len(tags)
+
+    @db_session
+    def save_tags(self, infohash: bytes, tags: Set[str]):
+        self.logger.debug(f'Save: {len(tags)} tags')
+        for tag in tags:
+            operation = TagOperation(
+                infohash=infohash,
+                operation=TagOperationEnum.ADD,
+                clock=CLOCK_START_VALUE,
+                creator_public_key=PUBLIC_KEY_FOR_AUTO_GENERATED_TAGS,
+                tag=tag
+            )
+            # we want auto generated operation to act like a normal operation
+            # therefore we use 2 as a `counter_increment` to immediately pass
+            # SHOW_THRESHOLD
+            self.db.add_auto_generated_tag_operation(operation=operation)
+
+    def _schedule_new_process_batch_round(self):
```

Comment:
**Question/issue:** It seems that the tag processor is a background task that is continuously running and periodically checks if there are new torrents available in the database to process. I think, however, that we can consider the tag processor as a one-shot task for now, mostly since the set of rules is fixed, and because new incoming torrents will be processed immediately. When we add additional rules through a new release, we can increase the version number of the processor and run the processor again on the database.

When we are also going to include rules created by users themselves, then we indeed need a continuous process. But since the rules are static at the moment, I would opt for running the processor once and when it has processed all torrents in the database, its task is done.

If we want to keep it continuously running, then I agree with @kozlovsky 's comment regarding the increase of the interval/batch size.

Reply:
Yes, agree, this was a feature engeneering from my side. I'll remove this logic :)

Reply:
Removed